# HW 3

Name: ChungHa Lee
<br>
ID: 20232027

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
boston = pd.read_csv('./Dataset/BostonHousing.csv')
boston.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

        b  lstat  medv  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

In [3]:
# Check for the null data in dataframe
boston.isnull().sum()

crim       0
zn         0
indus      0
chas       0
nox        0
rm         0
age        0
dis        0
rad        0
tax        0
ptratio    0
b          0
lstat      0
medv       0
dtype: int64

### Q7-(a). Based on this data set, provide an estimate for the population mean of medv data. Call this estimate μˆ.

In [4]:
boston['medv'].describe()

count    506.000000
mean      22.532806
std        9.197104
min        5.000000
25%       17.025000
50%       21.200000
75%       25.000000
max       50.000000
Name: medv, dtype: float64

In [5]:
boston['medv'].mean()

22.532806324110677

In [6]:
μˆ = 22.532806

### Q7-(b). Provide an estimate of the standard error of μˆ. Interpret this result. 

In [7]:
boston['medv'].count()

506

In [8]:
boston['medv'].std()

9.197104087379818

Standard Error (SE) = Standard Deviation (SD) / √(Sample Size)

SE = 9.197104 / √(506)
<br>
SE = 0.408861

### Q7-(c). Now estimate the standard error of μˆ using the bootstrap. How does this compare to your answer from (b)?

In [9]:
from scipy.stats import sem

In [10]:
sem(boston['medv'])

0.40886114749753505

### Q7-(d). Based on your bootstrap estimate from (c), provide a 95% confidence interval for the mean of medv. Compare it to the results obtained by using Boston[‘medv’].std() and the two standard error rule.

In [11]:
import scipy.stats as stats

medv = boston['medv'].tolist() 

sample_mean = np.mean(medv)
standard_deviation = np.std(medv)
sample_size = len(medv)
standard_error = standard_deviation / np.sqrt(sample_size)

confidence_level = 0.95
alpha = 1 - confidence_level
z_score = stats.norm.ppf(1 - alpha / 2)

lower_bound = sample_mean - (z_score * standard_error)
upper_bound = sample_mean + (z_score * standard_error)

print(f"95% Confidence Interval for 'medv' mean: ({lower_bound:.2f}, {upper_bound:.2f})")


95% Confidence Interval for 'medv' mean: (21.73, 23.33)


In [12]:
medv = boston['medv'].tolist() 

sample_mean = np.mean(medv)
standard_deviation = np.std(medv)
sample_size = len(medv)
standard_error = standard_deviation / np.sqrt(sample_size)

lower_bound = sample_mean - 2 * standard_error
upper_bound = sample_mean + 2 * standard_error

print(f"Two Standard Error Confidence Interval for 'medv' mean: ({lower_bound:.2f}, {upper_bound:.2f})")

Two Standard Error Confidence Interval for 'medv' mean: (21.72, 23.35)


### Q7-(e) Based on this data set, provide an estimate μˆmed, for the median value of medv in the population.

In [13]:
medv = boston['medv'].tolist() 

population_median_estimate = np.median(medv)

print(f"Estimated Population Median (μ^med): {population_median_estimate:.2f}")

Estimated Population Median (μ^med): 21.20


### Q7-(f) We now would like to estimate the standard error of μˆmed. Unfortunately, there is no simple formula for computing the standard error of the median. Instead, estimate the standard error of the median using the bootstrap. Comment on your findings. 

In [14]:
medv = boston['medv'].tolist() 


num_samples = 1000


bootstrapped_medians = np.zeros(num_samples)


for i in range(num_samples):
    bootstrap_sample = np.random.choice(medv, size=len(medv), replace=True)
    bootstrap_median = np.median(bootstrap_sample)
    bootstrapped_medians[i] = bootstrap_median

standard_error_median = np.std(bootstrapped_medians)

print(f"Standard Error of the Median (Estimated): {standard_error_median:.2f}")

Standard Error of the Median (Estimated): 0.37


### Q7-(g) Based on this data set, provide an estimate for the tenth percentile of medv in Boston census tracts. Call this quantity μˆ0.1. (You can use the np.percentile() function. 

In [15]:
medv = boston['medv'].tolist() 

tenth_percentile_estimate = np.percentile(medv, 10)

print(f"Estimated Tenth Percentile (μ^0.1) of 'medv': {tenth_percentile_estimate:.2f}")

Estimated Tenth Percentile (μ^0.1) of 'medv': 12.75


### Q7-(h) Use the bootstrap to estimate the standard error of μˆ0.1. Comment on your findings.

In [16]:
medv = boston['medv'].tolist() 

num_samples = 1000

bootstrapped_tenth_percentiles = np.zeros(num_samples)

for i in range(num_samples):
    bootstrap_sample = np.random.choice(medv, size=len(medv), replace=True)

    bootstrap_tenth_percentile = np.percentile(bootstrap_sample, 10)
    
    bootstrapped_tenth_percentiles[i] = bootstrap_tenth_percentile

standard_error_tenth_percentile = np.std(bootstrapped_tenth_percentiles)

print(f"Standard Error of the Tenth Percentile (Estimated): {standard_error_tenth_percentile:.2f}")

Standard Error of the Tenth Percentile (Estimated): 0.51
